### DISCUSSION TITLE
How to create JWKS KeyStore in 3.5.x?

### DISCUSSION THREAD

**Discussion by ajmas** [panva/jose#135] [Q&A] [answered]
I am wanting to adapt the code located at https://github.com/panva/node-oidc-provider-example/blob/master/01-oidc-configured/generate-keys.js, which uses jose 2.0.3, to use jose 3.5.0, but I am getting stuck since keystore appears to not be part of the latest version of Jose?

Replicated here:

```js
const fs = require('fs');
const path = require('path');
const jose = require('jose');

const keystore = new jose.JWKS.KeyStore();

Promise.all([
  keystore.generate('RSA', 2048, { use: 'sig' }),
  keystore.generate('RSA', 2048, { use: 'enc' }),
  keystore.generate('EC', 'P-256', { use: 'sig' }),
  keystore.generate('EC', 'P-256', { use: 'enc' }),
  keystore.generate('OKP', 'Ed25519', { use: 'sig' }),
]).then(() => {
  fs.writeFileSync(path.resolve('src/jwks.json'), JSON.stringify(keystore.toJWKS(true), null, 2));
});
```

Digging around it seems that `generateKeyPair()` from `jose/util/generate_key_pair` could be  a possibility, but it doesn't seem to cover all the parameters.

Any help would be appreciated.

**Comment by panva** [comment, marked as answer]
KeyStore equivalent is not in v3.x, I have not yet made my mind about what form or shape it'll take. 

That being said you do not need a KeyStore to form a JWKSet object.

**Reply by krnbr** [reply]
Until then need to use v2?

**Comment by kiehnek** [comment]
I wrote myself a small utility function to get a JWK from a JWKSet. I just copied the[ `getKey` function](https://github.com/panva/jose/blob/b024f96c572fbafa32b45d4ccf2a8a8645a4b376/src/jwks/remote.ts#L95) which is used by `jose` in a Remote KeySet. 

To create a KeySet you can simply create an Object with an Array `keys`, which holds the JWKs.

```js
{
  keys: [
    JWK1, JWK2, JWK3
  ]
}
```

This is my utility function: https://gist.github.com/korki43/c28cab45a33a5795aca0a4b7d68eff56

**Comment by grahhnt** [comment]
For anyone coming from Google:

- Generate JWK using [`jose.generateKeyPair`](https://github.com/panva/jose/blob/main/docs/functions/key_generate_key_pair.generateKeyPair.md)
- Export JWK using [`jose.exportJWK`](https://github.com/panva/jose/blob/main/docs/functions/key_export.exportJWK.md)
